src/common/provider_base.py

```diff
@@ -6,9 +6,81 @@
 across different broker APIs.
 """
 
-# TODO: Implement abstract base classes
-# - BaseDataProvider for market data streaming
-# - BaseTradeProvider for order execution
-# - Common error handling and retry logic
-# - Rate limiting and connection management
-# - Authentication and session management
\ No newline at end of file
+from abc import ABC, abstractmethod
+from typing import AsyncIterator, List
+from decimal import Decimal
+from .models import TradeTick, MetricUpdate, OrderAck, Position
+
+
+class DataProvider(ABC):
+    """Abstract base class for market data providers."""
+    
+    @abstractmethod
+    async def connect(self) -> None:
+        """Establish connection to data source."""
+        pass
+    
+    @abstractmethod
+    async def subscribe_trades(self, symbols: List[str]) -> None:
+        """Subscribe to trade data for given symbols."""
+        pass
+    
+    @abstractmethod  
+    async def subscribe_metrics(self, symbols: List[str]) -> None:
+        """Subscribe to metrics (mark-price, funding, liquidations)."""
+        pass
+    
+    @abstractmethod
+    async def iter_ticks(self) -> AsyncIterator[TradeTick]:
+        """Async iterator yielding trade ticks."""
+        pass
+    
+    @abstractmethod
+    async def iter_metrics(self) -> AsyncIterator[MetricUpdate]:
+        """Async iterator yielding metric updates."""
+        pass
+    
+    @abstractmethod
+    async def disconnect(self) -> None:
+        """Clean up and disconnect from data source."""
+        pass
+
+
+class TradeProvider(ABC):
```

Comment:
how are we modeling web socket vs REST? most brokerages and data providers support both. 
we need to model them correctly